crates/server/src/server/h3_handler.rs

```
@@ -0,0 +1,144 @@
+        Protocol, ResponseInfo,
+    },
+};
+
+pub(crate) async fn h3_handler<T>(
+    handler: Arc<T>,
+    mut connection: H3Connection,
+    src_addr: SocketAddr,
+    _dns_hostname: Option<Arc<str>>,
+    shutdown: Watch,
+) -> Result<(), ProtoError>
+where
+    T: RequestHandler,
+{
+    // TODO: we should make this configurable
+    let mut max_requests = 100u32;
+
+    // Accept all inbound requests sent over the connection.
+    loop {
+        let (_, mut stream) = tokio::select! {
+            result = connection.accept() => match result {
+                Some(Ok(next_request)) => next_request,
+                Some(Err(err)) => {
+                    warn!("error accepting request {}: {}", src_addr, err);
+                    return Err(err);
+                }
+                None => {
+                    break;
+                }
+            },
+            _ = shutdown.clone().signaled() => {
+                // A graceful shutdown was initiated.
+                break;
+            },
+        };
+
+        let request = match stream
+            .recv_data()
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 stream receive data failed: {e}")))?
+        {
+            Some(mut request) => request.copy_to_bytes(request.remaining()),
+            None => {
+                warn!("h3 stream received no data");
+                continue;
+            }
+        };
+
+        debug!(
+            "Received bytes {} from {src_addr} {request:?}",
+            request.remaining()
+        );
+        let handler = handler.clone();
+        let stream = Arc::new(Mutex::new(stream));
+        let responder = H3ResponseHandle(stream.clone());
+
+        handle_request(request, src_addr, handler, responder).await;
+
+        max_requests -= 1;
+        if max_requests == 0 {
+            warn!("exceeded request count, shutting down h3 conn: {src_addr}");
```

Same here, I would have said this should return an error instead of just a warning.

---

I'm trying to remember my exact reasoning on the return of Err or not here. I think it was that shutting down the connection is not an error condition, just normal behavior. But we could change that logic...

---

No, that makes sense to me actually.
I guess it would be nice to notify users if a protection mechanism was triggered, but that has to be changed in the whole library, not just here.